packages/guides-restructured-text/src/RestructuredText/Parser/InlineParser.php

```diff
@@ -8,34 +8,53 @@
 use phpDocumentor\Guides\Nodes\Inline\PlainTextInlineNode;
 use phpDocumentor\Guides\Nodes\InlineCompoundNode;
 use phpDocumentor\Guides\RestructuredText\Parser\Productions\InlineRules\InlineRule;
+use phpDocumentor\Guides\RestructuredText\Parser\Productions\InlineRules\MatchCachable;
 
 use function usort;
 
 class InlineParser
 {
     /** @var InlineRule[] */
     private array $rules;
+    private $lexer;
+
+    /** @var InlineRule[] */
+    private array $cachedRules = [];
 
     /** @param iterable<InlineRule> $inlineRules */
     public function __construct(iterable $inlineRules)
     {
+        $this->lexer = new InlineLexer();
         $this->rules = [...$inlineRules];
         usort($this->rules, static fn (InlineRule $a, InlineRule $b): int => $a->getPriority() > $b->getPriority() ? -1 : 1);
     }
 
     public function parse(string $content, BlockContext $blockContext): InlineCompoundNode
     {
-        $lexer = new InlineLexer();
-        $lexer->setInput($content);
-        $lexer->moveNext();
-        $lexer->moveNext();
+        $this->lexer->setInput($content);
+        $this->lexer->moveNext();
+        $this->lexer->moveNext();
         $nodes = [];
         $previous = null;
-        while ($lexer->token !== null) {
-            foreach ($this->rules as $inlineRule) {
+        while ($this->lexer->token !== null) {
+            if (isset($this->cachedRules[$this->lexer->token->type])) {
+                $node = $this->cachedRules[$this->lexer->token->type]->apply($blockContext, $this->lexer);
+                if ($node !== null) {
+                    $nodes[] = $node;
+                    $previous = $node;
+                    continue;
+                }
+            }
+
+            foreach ($this->rules as $key => $inlineRule) {
                 $node = null;
-                if ($inlineRule->applies($lexer)) {
-                    $node = $inlineRule->apply($blockContext, $lexer);
+                if ($inlineRule->applies($this->lexer)) {
+                    if ($inlineRule instanceof MatchCachable && $inlineRule->isCacheable()) {
+                        $this->cachedRules[$this->lexer->token->type] = $inlineRule;
+                        unset($this->rules[$key]);
+                    }
```

Comment:
Because we can remove the cached rules from the rules array we can remove about 1M calls to the parser rules. The inline rules are matched very often as very word and sometimes char can be a token. This foreach takes about 30% of the parsing time. 

![image](https://github.com/phpDocumentor/guides/assets/1060433/52c2c1e5-cbc9-4e85-a434-18a64e40b65f)